File: trabajo 12/heap_sort.cs
using System;

class HeapSortExample
{
    // Función para hacer heapify de un subárbol con raíz en el índice i
    static void Heapify(int[] arr, int n, int i)
    {
        int largest = i;       // Inicializar el más grande como raíz
        int left = 2 * i + 1;  // Hijo izquierdo
        int right = 2 * i + 2; // Hijo derecho

        // Si el hijo izquierdo es mayor que la raíz
        if (left < n && arr[left] > arr[largest])
            largest = left;

        // Si el hijo derecho es mayor que el más grande hasta ahora
        if (right < n && arr[right] > arr[largest])
            largest = right;

        // Si el mayor no es la raíz
        if (largest != i)
        {
            // Intercambiar
            int temp = arr[i];
            arr[i] = arr[largest];
            arr[largest] = temp;

            // Recursivamente hacer heapify al subárbol afectado
            Heapify(arr, n, largest);
        }
    }

    // Función principal para ordenar el arreglo usando Heap Sort
    static void HeapSort(int[] arr)
    {
        int n = arr.Length;

        // Construir Max Heap
        for (int i = n / 2 - 1; i >= 0; i--)
            Heapify(arr, n, i);

        // Extraer elementos uno por uno
        for (int i = n - 1; i > 0; i--)
        {
            // Mover la raíz al final
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;

            // Aplicar heapify al heap reducido
            Heapify(arr, i, 0);
        }
    }

    // Función para imprimir el arreglo
    static void PrintArray(int[] arr)
    {
        foreach (int num in arr)
            Console.Write(num + " ");
        Console.WriteLine();
    }

    // Método principal
    static void Main()
    {
        int[] arr = { 10, 20, 25, 5, 8, 7 };

        Console.WriteLine("Arreglo desordenado:");
        PrintArray(arr);

        HeapSort(arr);

        Console.WriteLine("Arreglo ordenado:");
        PrintArray(arr);
    }
}

File: trabajo 12/heap_sort.js
// Función para hacer heapify en un subárbol con raíz en el índice i
function heapify(arr, n, i) {
    let largest = i;
    const left = 2 * i + 1;
    const right = 2 * i + 2;

    // Si el hijo izquierdo es mayor que la raíz
    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    // Si el hijo derecho es mayor que el más grande hasta ahora
    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    // Si el mayor no es la raíz
    if (largest !== i) {
        [arr[i], arr[largest]] = [arr[largest], arr[i]]; // Intercambiar
        heapify(arr, n, largest); // Heapify recursivo
    }
}

// Función principal de Heap Sort
function heapSort(arr) {
    const n = arr.length;

    // Construir Max Heap
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }

    // Extraer elementos uno por uno
    for (let i = n - 1; i > 0; i--) {
        [arr[0], arr[i]] = [arr[i], arr[0]]; // Mover raíz al final
        heapify(arr, i, 0); // Heapify al heap reducido
    }
}

// Función para imprimir el arreglo
function printArray(arr) {
    console.log(arr.join(" "));
}

// Programa principal
const arr = [10, 20, 25, 5, 8, 7];

console.log("Arreglo desordenado:");
printArray(arr);

heapSort(arr);

console.log("Arreglo ordenado:");
printArray(arr);
